### PR TITLE
[SPARK-32963][SQL] empty string should be consistent for schema name in SparkGetSchemasOperation

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?
This PR makes the empty string for schema name pattern match the global temp view as same as it works for other databases.

This PR also add new tests to covering different kinds of wildcards to verify the SparkGetSchemasOperation

### Why are the changes needed?

When the schema name is empty string, it is considered as ".*" and can match all databases in the catalog.
But when it can not match the global temp view as it is not converted to ".*"

### Does this PR introduce _any_ user-facing change?

yes , JDBC operation like `statement.getConnection.getMetaData..getSchemas(null, "")` now also provides the global temp view in the result set.

### How was this patch tested?

new tests

### CODE REVIEW

**Comment by yaooqinn** [comment]
cc @cloud-fan @juliuszsompolski @maropu thanks~

**Comment by juliuszsompolski** [comment]
@bogdanghit

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkMetadataOperationSuite.scala [hunk 1]
```diff
@@ -28,23 +30,40 @@ class SparkMetadataOperationSuite extends HiveThriftJdbcTest {
 
   test("Spark's own GetSchemasOperation(SparkGetSchemasOperation)") {
```
Comment (cloud-fan) [inline sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkMetadataOperationSuite.scala:31]: Does this operation work for permanent views?

Comment (yaooqinn) [reply sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkMetadataOperationSuite.scala:31]: this is DB layer API, we need to fix other type of operation too

**Comment by cloud-fan** [comment]
Does the  `SHOW DATABASES` command list the "glob_temp" db？ I'm fine if they are consistent.

**Comment by yaooqinn** [comment]
> Does the `SHOW DATABASES` command list the "glob_temp" db？ I'm fine if they are consistent.

No. BUT they are not the same thing.

```java
if ((pattern == null) || pattern.isEmpty()) {
      return convertPattern("%", true);
} 
```
```java
if (pattern == null) {
      return ".*";
} 
```

the two Hive APIs we call here to convert "" or null to wildcards are different.

**Comment by yaooqinn** [comment]
> Does the `SHOW DATABASES` command list the "glob_temp" db？ I'm fine if they are consistent.

There are plenty of diffs between `SHOW XXX` commands and the thrift meta operations, e.g. the former only accepts '*' '.' as wildcards but the latter accepts '%' and '_' too

File: sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkGetSchemasOperation.scala [hunk 2]
```diff
@@ -77,7 +77,8 @@ private[hive] class SparkGetSchemasOperation(
 
       val globalTempViewDb = sqlContext.sessionState.catalog.globalTempViewManager.database
       val databasePattern = Pattern.compile(CLIServiceUtils.patternToRegex(schemaName))
-      if (databasePattern.matcher(globalTempViewDb).matches()) {
+      if (schemaName == null || schemaName.isEmpty ||
```
Comment (bogdanghit) [inline sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkGetSchemasOperation.scala:80]: Won't this break upstream client applications that depend on this behavior?

Comment (juliuszsompolski) [reply sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkGetSchemasOperation.scala:80]: https://docs.oracle.com/javase/7/docs/api/java/sql/DatabaseMetaData.html#getSchemas(java.lang.String,%20java.lang.String)
`schemaPattern - a schema name; must match the schema name as it is stored in the database; null means schema name should not be used to narrow down the search.`
This doc doesn't mention empty string, but if it's treated as a pattern, it should default to empty string not matching anything.
schemaName == null is already handled to match everything in patternToRegex:
```
  public static String patternToRegex(String pattern) {
    if (pattern == null) {
      return ".*";
    } else {
```

So the current behaviour seems to be consistent with JDBC documentation?

Comment (juliuszsompolski) [reply sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkGetSchemasOperation.scala:80]: https://docs.oracle.com/javase/7/docs/api/java/sql/DatabaseMetaData.html#getTables(java.lang.String,%20java.lang.String,%20java.lang.String,%20java.lang.String[])
`schemaPattern - a schema name pattern; must match the schema name as it is stored in the database; "" retrieves those without a schema; null means that the schema name should not be used to narrow the search`
The behaviour for getTables treats "" as no schema (e.g. local temp views), not all schemas, so it seems consistent that getSchemas wouldn't treat "" as "all schemas".

Comment (yaooqinn) [reply sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkGetSchemasOperation.scala:80]: Thanks @juliuszsompolski 

The java doc makes sense to me.

FYI, the current behavior in hive treats "" as wildcard,

```java
  /**
   * Convert wildchars and escape sequence of schema pattern from JDBC format to datanucleous/regex
   * The schema pattern treats empty string also as wildchar
   */

```
https://github.com/apache/hive/blob/14e3f19d027858a6f36ea6d41a3e8e45625f91bf/service/src/java/org/apache/hive/service/cli/operation/MetadataOperation.java#L75

We followed the hive behavior so far after we inlined the hive codes and before we made our own meta operations.

But now, "" is treated as wildcard for all schemas except ”global_temp“.

I guess we should either make it not a wildcard at all as JDBC API doc says or make it as same as null just like hive

Comment (cloud-fan) [reply sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkGetSchemasOperation.scala:80]: > we should either make it not a wildcard at all as JDBC API doc says'

Is it really possible? We call the hive API `listDatabases` to do it, and we can't change the hive metastore behavior.

Comment (juliuszsompolski) [reply sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkGetSchemasOperation.scala:80]: @cloud-fan 
this behaviour is changed by convertSchemaPattern in https://github.com/apache/hive/blob/14e3f19d027858a6f36ea6d41a3e8e45625f91bf/service/src/java/org/apache/hive/service/cli/operation/MetadataOperation.java#L75
```
    if ((pattern == null) || pattern.isEmpty()) {
      return convertPattern("%", true);
    } else {
      return convertPattern(pattern, true);
    }
```
it explicitly turns "" into a "%" pattern.
Then we use that to call listDatabases:
```
      val schemaPattern = convertSchemaPattern(schemaName)
      sqlContext.sessionState.catalog.listDatabases(schemaPattern).foreach { dbName =>
        rowSet.addRow(Array[AnyRef](dbName, DEFAULT_HIVE_CATALOG))
      }
```
but then for globalTempViews, we are not using the converted schemaPattern, but schemaName, which will not treat "" as wildcard.
```
      val globalTempViewDb = sqlContext.sessionState.catalog.globalTempViewManager.database
      val databasePattern = Pattern.compile(CLIServiceUtils.patternToRegex(schemaName))
      if (databasePattern.matcher(globalTempViewDb).matches()) {
```

described by @yaooqinn like that, it does seem to be inconsistent behaviour and just a omission bug: we should treat the pattern for global temp views the same way... JDBC standard would say that it should not be a wildcard, but Hive seems to be treating it as wildcard.
Treating it as empty would be consistent with JDBC standard. Treating it as a wildcard would be a smaller behaviour change, and consistent with Hive.
@bogdanghit @wangyum @AngersZhuuuu what do you think?

Comment (cloud-fan) [reply sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkGetSchemasOperation.scala:80]: Given that changing from Hive behavior to JDBC standard is a very breaking change ('' is not wildcard anymore), I'd prefer to fix this small behavior for global temp view here.

Comment (bogdanghit) [reply sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkGetSchemasOperation.scala:80]: sounds good. thanks a lot for the context added here @juliuszsompolski.

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
retest this please

**Comment by yaooqinn** [comment]
retest this please

**Comment by cloud-fan** [comment]
retest this please

**Comment by maropu** [comment]
retest this please

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**